File: web/src/app/modules/recouvrement/recouvrement.service.ts
import { Injectable } from "@angular/core";
import { HttpClient, HttpParams } from "@angular/common/http";
import { Observable } from "rxjs";
import { map } from "rxjs/operators";
import { ApiResponse, PageResponse } from "@core/models/api-response.model";
import {
  DossierRecouvrementResponse,
  ActionRecouvrementResponse,
  AccordReechelonnementResponse,
  OuvrirDossierRequest,
  AjouterActionRequest,
  EscaladerDossierRequest,
  AccordReechelonnementRequest,
  RecouvrementPhase,
} from "@core/models/recouvrement.model";

@Injectable({ providedIn: "root" })
export class RecouvrementService {
  private readonly API = "/api/recouvrement";

  constructor(private http: HttpClient) {}

  // ── Dossiers ────────────────────────────────────────────────────────────────

  ouvrirDossier(
    req: OuvrirDossierRequest,
  ): Observable<DossierRecouvrementResponse> {
    return this.http
      .post<
        ApiResponse<DossierRecouvrementResponse>
      >(`${this.API}/dossiers`, req)
      .pipe(map((r) => r.data));
  }

  listDossiers(
    phase?: RecouvrementPhase,
    clos?: boolean,
    page = 0,
    size = 20,
  ): Observable<PageResponse<DossierRecouvrementResponse>> {
    let params = new HttpParams().set("page", page).set("size", size);
    if (phase) params = params.set("phase", phase);
    if (clos != null) params = params.set("clos", String(clos));
    return this.http
      .get<
        ApiResponse<PageResponse<DossierRecouvrementResponse>>
      >(`${this.API}/dossiers`, { params })
      .pipe(map((r) => r.data));
  }

  getDossier(id: number): Observable<DossierRecouvrementResponse> {
    return this.http
      .get<
        ApiResponse<DossierRecouvrementResponse>
      >(`${this.API}/dossiers/${id}`)
      .pipe(map((r) => r.data));
  }

  escalader(
    id: number,
    req: EscaladerDossierRequest,
  ): Observable<DossierRecouvrementResponse> {
    return this.http
      .put<
        ApiResponse<DossierRecouvrementResponse>
      >(`${this.API}/dossiers/${id}/escalader`, req)
      .pipe(map((r) => r.data));
  }

  clore(id: number, motif = ""): Observable<DossierRecouvrementResponse> {
    const params = new HttpParams().set("motif", motif);
    return this.http
      .put<
        ApiResponse<DossierRecouvrementResponse>
      >(`${this.API}/dossiers/${id}/clore`, null, { params })
      .pipe(map((r) => r.data));
  }

  // ── Actions ─────────────────────────────────────────────────────────────────

  ajouterAction(
    dossierId: number,
    req: AjouterActionRequest,
  ): Observable<ActionRecouvrementResponse> {
    return this.http
      .post<
        ApiResponse<ActionRecouvrementResponse>
      >(`${this.API}/dossiers/${dossierId}/actions`, req)
      .pipe(map((r) => r.data));
  }

  getActions(dossierId: number): Observable<ActionRecouvrementResponse[]> {
    return this.http
      .get<
        ApiResponse<ActionRecouvrementResponse[]>
      >(`${this.API}/dossiers/${dossierId}/actions`)
      .pipe(map((r) => r.data));
  }

  // ── Accords ─────────────────────────────────────────────────────────────────

  creerAccord(
    dossierId: number,
    req: AccordReechelonnementRequest,
  ): Observable<AccordReechelonnementResponse> {
    return this.http
      .post<
        ApiResponse<AccordReechelonnementResponse>
      >(`${this.API}/dossiers/${dossierId}/accords`, req)
      .pipe(map((r) => r.data));
  }

  getAccords(dossierId: number): Observable<AccordReechelonnementResponse[]> {
    return this.http
      .get<
        ApiResponse<AccordReechelonnementResponse[]>
      >(`${this.API}/dossiers/${dossierId}/accords`)
      .pipe(map((r) => r.data));
  }
}

File: mobile/lib/core/services/recouvrement_service.dart
import '../models/recouvrement.dart';
import 'api_service.dart';

class PagedRecouvrement {
  final List<DossierRecouvrement> content;
  final int totalElements;
  const PagedRecouvrement({required this.content, required this.totalElements});
}

class RecouvrementService {
  final ApiService _api;
  RecouvrementService(this._api);

  Future<PagedRecouvrement> listDossiers({
    String? phase,
    bool? clos,
    int page = 0,
    int size = 20,
  }) async {
    final params = <String, dynamic>{'page': page, 'size': size};
    if (phase != null) params['phase'] = phase;
    if (clos != null) params['clos'] = clos.toString();

    final resp = await _api.get<Map<String, dynamic>>(
      '/api/v1/recouvrement/dossiers',
      queryParameters: params,
      fromJson: (d) => d as Map<String, dynamic>,
    );
    final data = resp['data'] as Map<String, dynamic>?;
    final items = (data?['content'] as List<dynamic>? ?? [])
        .map((e) => DossierRecouvrement.fromJson(e as Map<String, dynamic>))
        .toList();
    return PagedRecouvrement(
      content: items,
      totalElements: (data?['totalElements'] as num?)?.toInt() ?? 0,
    );
  }

  Future<DossierRecouvrement> getDossier(int id) async {
    final resp = await _api.get<Map<String, dynamic>>(
      '/api/v1/recouvrement/dossiers/$id',
      fromJson: (d) => d as Map<String, dynamic>,
    );
    return DossierRecouvrement.fromJson(resp['data'] as Map<String, dynamic>);
  }

  Future<List<ActionRecouvrement>> getActions(int dossierId) async {
    final resp = await _api.get<Map<String, dynamic>>(
      '/api/v1/recouvrement/dossiers/$dossierId/actions',
      fromJson: (d) => d as Map<String, dynamic>,
    );
    final list = resp['data'] as List<dynamic>? ?? [];
    return list.map((e) => ActionRecouvrement.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<ActionRecouvrement> ajouterAction(int dossierId, Map<String, dynamic> body) async {
    final resp = await _api.post<Map<String, dynamic>>(
      '/api/v1/recouvrement/dossiers/$dossierId/actions',
      data: body,
      fromJson: (d) => d as Map<String, dynamic>,
    );
    return ActionRecouvrement.fromJson(resp['data'] as Map<String, dynamic>);
  }

  Future<DossierRecouvrement> escalader(int id, String nouvellePhase, {String? motif}) async {
    final resp = await _api.put<Map<String, dynamic>>(
      '/api/v1/recouvrement/dossiers/$id/escalader',
      data: {'nouvellePhase': nouvellePhase, if (motif != null) 'motif': motif},
      fromJson: (d) => d as Map<String, dynamic>,
    );
    return DossierRecouvrement.fromJson(resp['data'] as Map<String, dynamic>);
  }
}
